kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java

```diff
@@ -165,7 +165,12 @@ public LogReplay(
                 loadTableProtocolAndMetadata(
                     engine, logSegment, newerSnapshotHint, logSegment.getVersion()));
     // Lazy loading of domain metadata only when needed
-    this.activeDomainMetadataMap = new Lazy<>(() -> loadDomainMetadataMap(engine));
+    this.activeDomainMetadataMap =
```

Comment:
I think we can combine the loadProtocolAndMetadata and loadMetadata into one to avoid double log replay. In most  happy paths, we get P&M and DMs from the CRC file. When CRC is missing, we need to do log replay, and the perf is expected to be bad. Why do two log replays? Instead, do one log replay and be smart about the case where the table has no DM Just don't look for DMs in the log replay and exit as soon as the P&M is found.

There is one case where the DM is not written to the CRC file: in this case we can do a log replay, but we don't need a new method, just use the one log replay method created for loading P&M and DMs.

Reply:
Discussed offline, we will do the replay combination in a separate PR. (We need to also revisit the snapshot hint logic today, which adds lots of code complexity here)